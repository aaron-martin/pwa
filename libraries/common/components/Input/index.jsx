/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * A component that takes care of rendering and validation of input fields.
 * This component has no styling and should not be used directly.
 * You may want to use an appropriate form field component from the template instead.
 */
class Input extends Component {

  static propTypes = {
    className: PropTypes.string,
    disabled: PropTypes.bool,
    onChange: PropTypes.func,
    onFocusChange: PropTypes.func,
    onSanitize: PropTypes.func,
    onValidate: PropTypes.func,
    password: PropTypes.bool,
    setRef: PropTypes.func,
    type: PropTypes.string,
    validateOnBlur: PropTypes.bool,
    value: PropTypes.string,
  };

  static defaultProps = {
    className: '',
    disabled: false,
    onChange: () => {},
    onFocusChange: () => {},
    onSanitize: value => value,
    onValidate: () => true,
    password: false,
    setRef: () => {},
    type: 'text',
    validateOnBlur: true,
    value: '',
  };

  /**
   * Creates a new input component.
   * @param {Object} props The component properties.
   */
  constructor(props) {
    super(props);

    // Initially sanitize the value.
    const sanitizedValue = this.props.onSanitize(this.props.value || '');

    this.state = {
      value: sanitizedValue,
      isValid: this.props.onValidate(sanitizedValue, true),
      isFocused: false,
    };
  }

  /**
   * Initially trigger onChange() to set the initial value.
   */
  componentDidMount() {
    this.props.onChange(this.state.value);
  }

  /**
   * Updates the value of the input field if the props has been modified.
   * @param {Object} nextProps The new properties.
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.value !== this.props.value) {
      /**
       * Only set the state value if the value prop has been changed,
       * otherwise use the current input state.
       */
      const sanitizedValue = this.props.onSanitize(nextProps.value || '');
      this.updateValue(sanitizedValue);
    }
  }

  /**
   * Internal focus event handler.
   */
  handleFocus = () => {
    this.setState({ isFocused: true });
    this.props.onFocusChange(true);
  };

  /**
   * Internal blur event handler.
   */
  handleBlur = () => {
    const newState = {
      isFocused: false,
    };

    if (this.props.validateOnBlur) {
      // Validate the value.
      newState.isValid = this.props.onValidate(this.value, false);
    }

    this.setState(newState);

    this.props.onFocusChange(false);
  };

  /**
   * Internal change event handler.
   * @param {Object} event The event object.
   */
  handleChange = (event) => {
    // Sanitize the input value.
    const sanitizedValue = this.props.onSanitize(event.target.value || '');

    // Update the state.
    this.updateValue(sanitizedValue);

    // Emit an event.
    this.props.onChange(sanitizedValue);
  };

  /**
   * Updates and validates the internal state value of the input field.
   * @param {string} newValue The new value.
   */
  updateValue(newValue) {
    const newState = {
      value: newValue,
    };

    if (!this.props.validateOnBlur) {
      // Validate while typing.
      newState.isValid = this.props.onValidate(newValue, false);
    }

    this.setState(newState);
  }

  /**
   * @returns {boolean} Whether the current value of the input field is valid.
   */
  get isValid() {
    return this.state.isValid;
  }

  /**
   * @return {boolean} Whether the input field is focused.
   */
  get isFocused() {
    return this.state.isFocused;
  }

  /**
   * @returns {string} The current value of the input field.
   */
  get value() {
    return this.state.value;
  }

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    const { className, password } = this.props;
    const type = password ? 'password' : this.props.type;
    const { value } = this.state;

    return (
      <input
        ref={this.props.setRef}
        className={className}
        type={type}
        value={value}
        onChange={this.handleChange}
        onFocus={this.handleFocus}
        onBlur={this.handleBlur}
        disabled={this.props.disabled}
      />
    );
  }
}

export default Input;
