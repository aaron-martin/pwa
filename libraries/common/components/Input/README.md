Input
---
This component will render a plain input field. You may consider to use a 
styled template component instead.

## Props

### className (optional)

_Type_: `className`<br>

Any additional style classes that should be applied to the component.

### onChange (optional)

_Type_: `function (value)`<br>

An optional callback that is triggered when the value of the input field
has been changed.

### onFocusChange (optional)

_Type_: `function (isFocused)`<br>

An optional callback that is called when the input field gains or loses
the current focus.

### onSanitize (optional)

_Type_: `function (value)`<br>

An optional callback that can be used to transform the value entered by the
user on the fly. The function is expected to return the sanitized value.

### onValidate (optional)

_Type_: `function (value, isInitial)`<br>

An optional callback that can be used to validate the value entered by the
user on the fly. The function is expected to return a boolean indicating
whether the value is valid.

The second parameter, `isInitial` indicates whether this is the initial
value validation that is happening in the construction/pre-mount phase of
the component. You can use this information to bypass any default value or
take care of proper `setState()` usages. 

By default, input validation will only happen if the field is blurred
(lost its focus). See `validateOnBlur` for further reference.

### password (optional)

_Type_: `boolean`<br>
_Default_: `false`

If this prop is set, the component will be rendered as a password field.

### value (optional)

_Type_: `string`<br>

The initial value of the input field. If the value of this property is modified,
the current value entered by the user is overwritten.

### validateOnBlur (optional)

_Type_: `boolean`<br>
_Default_: `true`<br>

By default, input validation will only happen if the field is blurred
(lost its focus). To enable validation while the user is typing, this property
can be set to false.

## Usage:

```markup
import { Input } from 'Library/components';

// Makes the entered value lowercase.
doSanitize = (value) => value.toLowerCase();

// Only accept values that contain the word 'Hello'.
doValidate = (value) => value.indexOf('Hello') != -1;

<Input
    onValidate={doValidate}
    onSanitize={doSanitize}
    value="Hello, World"
/>
```


