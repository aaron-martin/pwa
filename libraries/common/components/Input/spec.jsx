/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import Input from './index';

describe('<Input />', () => {
  it('should render a simple input field', () => {
    const wrapper = mount(<Input />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input').length).toBe(1);
  });

  it('should render the input as password', () => {
    const wrapper = mount(<Input password />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input[type="password"]').length).toBe(1);
  });

  it('should render the input with a default value', () => {
    const wrapper = mount(<Input value="FooBar" />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input[value="FooBar"]').length).toBe(1);
  });

  it('should trigger the onChange callback', () => {
    const onChangeMock = jest.fn();

    const wrapper = mount(<Input onChange={onChangeMock} />);

    wrapper.find('input').simulate('change', { target: { value: 'a' } });

    expect(wrapper).toMatchSnapshot();
    expect(onChangeMock.mock.calls.length).toBe(2);
  });

  it('should receive the correct value while typing', () => {
    const wrapper = mount(<Input />);

    const input = wrapper.find('input');

    input.simulate('change', { target: { value: 'foobar' } });

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('SimpleInput').instance().value).toBe('foobar');
  });

  it('should sanitize the input', () => {
    const wrapper = mount(<Input onSanitize={value => value.toUpperCase()} />);

    const input = wrapper.find('input');

    input.simulate('change', { target: { value: 'foobar' } });

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('SimpleInput').instance().value).toBe('FOOBAR');
  });

  it('should validate the input', () => {
    const wrapper = mount(<Input onValidate={() => false} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('SimpleInput').instance().isValid).toBe(false);
  });

  it('should focus the input', () => {
    const onFocusMock = jest.fn();

    const wrapper = mount(<Input onFocusChange={onFocusMock} />);

    const input = wrapper.find('input');

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('SimpleInput').instance().isFocused).toBe(false);

    input.simulate('focus');

    expect(wrapper.find('SimpleInput').instance().isFocused).toBe(true);

    input.simulate('blur');
    expect(wrapper.find('SimpleInput').instance().isFocused).toBe(false);
  });

  it('should change the value on user input', () => {
    const wrapper = mount(<Input value="My initial value" />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('SimpleInput').instance().value).toBe('My initial value');

    const input = wrapper.find('input');
    input.simulate('change', { target: { value: 'foobar' } });

    expect(wrapper.find('SimpleInput').instance().value).toBe('foobar');
  });

  it('should render a multiline input with empty content and react on change', () => {
    const multiLineValue = `dfsdsdf
    sdfdsff
    dsf`;
    const wrapper = mount(<Input value="" multiLine />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('MultiLineInput').instance().value).toEqual('');
    wrapper.setProps({ value: multiLineValue });
    expect(wrapper.find('textarea').getDOMNode().innerHTML).toEqual(multiLineValue);
  });
});
