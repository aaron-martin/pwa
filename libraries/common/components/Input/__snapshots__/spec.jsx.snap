// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should change the value on user input 1`] = `
<Input
  className=""
  disabled={false}
  onChange={[Function]}
  onFocusChange={[Function]}
  onSanitize={[Function]}
  onValidate={[Function]}
  password={false}
  setRef={[Function]}
  type="text"
  validateOnBlur={true}
  value="My initial value"
>
  <input
    className=""
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="My initial value"
  />
</Input>
`;

exports[`<Input /> should focus the input 1`] = `
<Input
  className=""
  disabled={false}
  onChange={[Function]}
  onFocusChange={[Function]}
  onSanitize={[Function]}
  onValidate={[Function]}
  password={false}
  setRef={[Function]}
  type="text"
  validateOnBlur={true}
  value=""
>
  <input
    className=""
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
</Input>
`;

exports[`<Input /> should receive the correct value while typing 1`] = `
<Input
  className=""
  disabled={false}
  onChange={[Function]}
  onFocusChange={[Function]}
  onSanitize={[Function]}
  onValidate={[Function]}
  password={false}
  setRef={[Function]}
  type="text"
  validateOnBlur={true}
  value=""
>
  <input
    className=""
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="foobar"
  />
</Input>
`;

exports[`<Input /> should render a simple input field 1`] = `
<Input
  className=""
  disabled={false}
  onChange={[Function]}
  onFocusChange={[Function]}
  onSanitize={[Function]}
  onValidate={[Function]}
  password={false}
  setRef={[Function]}
  type="text"
  validateOnBlur={true}
  value=""
>
  <input
    className=""
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
</Input>
`;

exports[`<Input /> should render the input as password 1`] = `
<input
  className=""
  disabled={false}
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  type="password"
  value=""
/>
`;

exports[`<Input /> should render the input with a default value 1`] = `
<input
  className=""
  disabled={false}
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  type="text"
  value="FooBar"
/>
`;

exports[`<Input /> should sanitize the input 1`] = `
<Input
  className=""
  disabled={false}
  onChange={[Function]}
  onFocusChange={[Function]}
  onSanitize={[Function]}
  onValidate={[Function]}
  password={false}
  setRef={[Function]}
  type="text"
  validateOnBlur={true}
  value=""
>
  <input
    className=""
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="FOOBAR"
  />
</Input>
`;

exports[`<Input /> should trigger the onChange callback 1`] = `
<Input
  className=""
  disabled={false}
  onChange={[Function]}
  onFocusChange={[Function]}
  onSanitize={[Function]}
  onValidate={[Function]}
  password={false}
  setRef={[Function]}
  type="text"
  validateOnBlur={true}
  value=""
>
  <input
    className=""
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="a"
  />
</Input>
`;

exports[`<Input /> should validate the input 1`] = `
<input
  className=""
  disabled={false}
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  type="text"
  value=""
/>
`;
